examples/chef/esp32/main/main.cpp

```diff
@@ -0,0 +1,209 @@
+        {
+            ChipLogProgress(Shell, "IPv6 Server ready...");
+            chip::app::DnssdServer::Instance().StartServer();
+        }
+        else if (event->InternetConnectivityChange.IPv6 == kConnectivity_Lost)
+        {
+            ChipLogProgress(Shell, "Lost IPv6 connectivity...");
+        }
+
+        break;
+
+    case DeviceEventType::kSessionEstablished:
+        if (event->SessionEstablished.IsCommissioner)
+        {
+            ChipLogProgress(Shell, "Commissioner detected!");
+        }
+        break;
+
+    case DeviceEventType::kCHIPoBLEConnectionEstablished:
+        ChipLogProgress(Shell, "CHIPoBLE connection established");
+        break;
+
+    case DeviceEventType::kCHIPoBLEConnectionClosed:
+        ChipLogProgress(Shell, "CHIPoBLE disconnected");
+        break;
+
+    case DeviceEventType::kCommissioningComplete:
+        ChipLogProgress(Shell, "Commissioning complete");
+        break;
+
+    case DeviceEventType::kInterfaceIpAddressChanged:
+        if ((event->InterfaceIpAddressChanged.Type == InterfaceIpChangeType::kIpV4_Assigned) ||
+            (event->InterfaceIpAddressChanged.Type == InterfaceIpChangeType::kIpV6_Assigned))
+        {
+            // DNSSD server restart on any ip assignment: if link local ipv6 is configured, that
+            // will not trigger a 'internet connectivity change' as there is no internet
+            // connectivity. DNSSD still wants to refresh its listening interfaces to include the
+            // newly selected address.
+            chip::app::DnssdServer::Instance().StartServer();
+        }
+        break;
+    }
+
+    ChipLogProgress(Shell, "Current free heap: %zu\n", heap_caps_get_free_size(MALLOC_CAP_8BIT));
+}
+
+const char * TAG = "chef-app";
+
+#if CONFIG_HAVE_DISPLAY
+void printQRCode()
+{
+    std::string qrCodeText;
+
+    GetQRCode(qrCodeText, chip::RendezvousInformationFlags(CONFIG_RENDEZVOUS_MODE));
+
+    // Initialize the display device.
+    esp_err_t err = InitDisplay();
+    if (err != ESP_OK)
+    {
+        ChipLogError(Shell, "InitDisplay() failed: %s", esp_err_to_name(err));
+        return;
+    }
+
+    // Initialize the screen manager
+    ScreenManager::Init();
+
+    ESP_LOGI(TAG, "Opening QR code screen");
+    ESP_LOGI(TAG, "QR CODE Text: '%s'", qrCodeText.c_str());
+    ScreenManager::PushScreen(chip::Platform::New<QRCodeScreen>(qrCodeText));
+}
+#endif // CONFIG_HAVE_DISPLAY
+
+extern "C" void app_main(void)
+{
+    ESP_ERROR_CHECK(nvs_flash_init());
+    chip::Platform::MemoryInit();
+    chip::DeviceLayer::PlatformMgr().InitChipStack();
+    chip::DeviceLayer::PlatformMgr().StartEventLoopTask();
+
+#if CONFIG_ENABLE_CHIP_SHELL
+    int ret = Engine::Root().Init();
+    VerifyOrDie(ret == 0);
+#endif /* CONFIG_ENABLE_CHIP_SHELL */
+
+    // Network connectivity
+    // Note to integration: StartWiFiManagement does not exist on ESP32
+
+    // Start IM server
+    chip::Server::GetInstance().Init();
```

Comment:
Needs to be done on the chip thread ... something like `chip::DeviceLayer::PlatformMgr().ScheduleWork(...)`